polymer/foam-select.html

```diff
@@ -0,0 +1,138 @@
+<link rel="import" href="./bower_components/polymer/polymer.html">
+
+<dom-module id="foam-select">
+  <script>
+  Polymer({
+    is: 'foam-select',
+
+    properties: {
+      dao: Object,
+      data: {
+        type: Array,
+        notify: true
+      }
+    },
+
+    observers: [
+      '_onDataUpdate(data.*)'
+    ],
+
+    ready: function() {
+      this._ids = {};
+
+      this._sink = {
+        put: this._put.bind(this),
+        remove: this._remove.bind(this),
+        reset: this._reset.bind(this),
+        eof: function() {}
+      };
+
+      this.feedback_ = true;
+      this.data = [];
+      this.feedback_ = false;
+
+      this.dao = eval(this.dao);
+      this.dao.pipe(this._sink);
+    },
+
+    _put: function(obj) {
+      obj = obj.clone();
+      obj.propertyChange.sub(function(sub) {
+        sub.destroy();
+        this.dao.put(obj);
+      }.bind(this));
+
+      if ( typeof this._ids[obj.id] !== 'undefined' ) {
+        // Replace that user with this new one.
+        this.feedback_ = true;
+        this.splice('data', this._ids[obj.id], 1, obj);
+        this.feedback_ = false;
+        return;
+      }
+
+      this._ids[obj.id] = this.data.length;
+      this.feedback_ = true;
+      this.push('data', obj);
+      this.feedback_ = false;
+    },
+
+    _remove: function(obj) {
+      if ( typeof this._ids[obj.id] !== 'undefined' ) {
+        this.feedback_ = true;
+        this.splice('data', this._ids[obj.id], 1);
+        this.feedback_ = false;
+        delete this._ids[obj.id];
+      }
+    },
+
+    _reset: function() {
+      this.feedback_ = true;
+      this.data = [];
+      this._ids = {};
+      this.feedback_ = false;
+      this.dao.select(this._sink);
+    },
+
+    _onDataUpdate: function(e) {
+      if ( this.feedback_ ) return; // Do nothing if I made this change.
+      // Examine the splice for additions, removals and edits.
+      var splices = e.value.indexSplices;
+      if ( ! splices ) return;
+      for ( var i = 0 ; i < splices.length ; i++ ) {
+        var splice = splices[i];
+        var removedKeys = {};
+        var addedKeys = {};
+
+        for ( var j = 0 ; j < splice.addedCount ; j++ ) {
+          addedKeys[splice.object[splice.index + j].id] = true;
+        }
+        // Now everything that's in removed but not in addedKeys needs to be
+        // removed from the DAO.
+        for ( var j = 0 ; j < splice.removed.length ; j++ ) {
+          var obj = splice.removed[i];
+          if ( ! addedKeys[obj.id] ) {
+            delete this._ids[obj.id];
+            window.setTimeout(this._asyncRemove.bind(this, obj), 0);
```

Comment:
Design wise, I would move the setTimeout into the _asyncPut/_asyncRemove functions so you can just call this._asyncPut(obj) here.